atuin-client/src/history.rs

```diff
@@ -5,38 +5,72 @@ use serde::{Deserialize, Serialize};
 
 use atuin_common::utils::uuid_v7;
 
-// Any new fields MUST be Optional<>!
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, sqlx::FromRow)]
+mod builder;
+mod context;
+
+pub use context::Context;
+
+/// A marker type used to seal the `History` struct, preventing it from being constructed directly.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub(self) struct HistorySeal;
+
+/// Client-side history entry.
+///
+/// Client stores data unencrypted, and only encrypts it before sending to the server.
+///
+/// To create a new history entry, use one of the builders:
+/// - [`History::import()`] to import an entry from the shell history file
+/// - [`History::capture()`] to capture an entry via hook
+/// - [`History::from_db()`] to create an instance from the database entry
+///
+/// ### Caution
+/// Any new fields MUST be `Optional<T>` and marked with `#[serde(default)]` to ensure backwards
+/// compatibility with older clients.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
 pub struct History {
+    /// A client-generated ID, used to identify the entry when syncing.
+    ///
+    /// Stored as `client_id` in the database.
     pub id: String,
+    /// When the command was run.
     pub timestamp: chrono::DateTime<Utc>,
+    /// How long the command took to run.
     pub duration: i64,
+    /// The exit code of the command.
     pub exit: i64,
+    /// The command that was run.
     pub command: String,
+    /// The current working directory when the command was run.
     pub cwd: String,
+    /// The session ID, associated with a terminal session.
     pub session: String,
+    /// The hostname of the machine the command was run on.
     pub hostname: String,
+
+    /// Timestamp, which is set when the entry is deleted, allowing a soft delete.
+    #[serde(default)]
     pub deleted_at: Option<chrono::DateTime<Utc>>,
-}
 
-// Forgive me, for I have sinned
-// I need to replace rmp with something that is more backwards-compatible.
-// Protobuf, or maybe just json
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, sqlx::FromRow)]
-pub struct HistoryWithoutDelete {
-    pub id: String,
-    pub timestamp: chrono::DateTime<Utc>,
-    pub duration: i64,
-    pub exit: i64,
-    pub command: String,
-    pub cwd: String,
-    pub session: String,
-    pub hostname: String,
+    /// Interpreter (usually some kind of shell) that was used to run the command.
+    #[serde(default)]
+    pub interpreter: Option<String>,
```

Comment:
I've decided to make it a distinct parameter in the end, but that's my take, I'm totally fine with reconsidering it, given the arguments :)

For now it just felt more natural – for example, it makes sense to enable interpreter capture by default, but disable env vars capture.